### PR TITLE
prov/efa: fix a bug in zero copy receive

### PR DESCRIPTION
**Pull request by wzamazon**
This PR contains a series of patches that fixed a bug in the zero copy receive model.

### CODE REVIEW

File: prov/efa/src/rxr/rxr_pkt_type_req.c [hunk 1]
```diff
@@ -997,6 +997,71 @@ ssize_t rxr_pkt_proc_matched_medium_rtm(struct rxr_ep *ep,
...
+		rx_entry->cq_entry.len = 0;
+	} else {
+		assert(rx_entry->cq_entry.buf == pkt_entry->pkt - sizeof(struct rxr_pkt_entry));
+		rx_entry->cq_entry.len = pkt_entry->pkt_size + sizeof(struct rxr_pkt_entry);
+	}
+
+	rxr_cq_write_rx_completion(ep, rx_entry);
+	rxr_release_rx_entry(ep, rx_entry);
+
+	/* no need to release packet enry because it is
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_pkt_type_req.c:1060]: enry -> entry

File: prov/efa/src/rxr/rxr_pkt_cmd.c [hunk 2]
```diff
@@ -823,60 +823,17 @@ fi_addr_t rxr_pkt_insert_addr(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry
 	return rdm_addr;
 }
 
-void rxr_pkt_handle_recv_completion(struct rxr_ep *ep,
-				    struct rxr_pkt_entry *pkt_entry)
+/**
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_pkt_cmd.c:826]: Typo in commit ```prov/efa: handle the case user posted buffer matched with non-rtm packet```

to recive application data. -> to receive application data.

File: prov/efa/src/rxr/rxr_ep.c [hunk 3]
```diff
@@ -302,6 +302,37 @@ int rxr_ep_post_prov_buf(struct rxr_ep *ep, uint64_t flags, enum rxr_lower_ep_ty
 	return 0;
 }
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_ep.c:304]: Typo in commit ```prov/efa: adjust internal buffer posting in zero copy receive mode```

If application posted recevie buffer -> If the application posted a receive buffer,

File: prov/efa/src/rxr/rxr_ep.c [hunk 4]
```diff
@@ -1359,33 +1364,77 @@ struct fi_ops_cm rxr_ep_cm = {
 	.join = fi_no_join,
 };
 
-static inline int rxr_ep_bulk_post_recv(struct rxr_ep *ep)
+/**
+ * @brief post internal receive bufers for progress engine.
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_ep.c:1368]: bufers -> buffers

File: prov/efa/src/rxr/rxr_ep.c [hunk 5]
```diff
@@ -1359,33 +1364,77 @@ struct fi_ops_cm rxr_ep_cm = {
...
+ *
+ * It is more efficient to post multiple receive buffers
+ * to device at once than to post each receive buffer
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_ep.c:1371]: to device -> to the device

File: prov/efa/src/rxr/rxr_ep.c [hunk 6]
```diff
@@ -302,6 +302,37 @@ int rxr_ep_post_prov_buf(struct rxr_ep *ep, uint64_t flags, enum rxr_lower_ep_ty
 	return 0;
 }
 
+/**
+ * @brief post multiple internal receive buffers to device
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_ep.c:306]: change multiple to something else, since this number can be 1

File: prov/efa/src/rxr/rxr_msg.c [hunk 7]
```diff
@@ -1030,7 +1030,7 @@ ssize_t rxr_msg_generic_recv(struct fid_ep *ep, const struct fi_msg *msg,
 	 * applicable to the zero-copy path where unexpected messages are not
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_msg.c:1030]: Need to change the comment

File: prov/efa/src/efa.h [hunk 8]
```diff
@@ -544,6 +544,34 @@ bool efa_both_support_rdma_read(struct rxr_ep *ep, struct rdm_peer *peer)
 	       (peer->is_self || efa_peer_support_rdma_read(peer));
 }
 
+/**
+ * @brief determine whether a peer need the endpoint to include raw address
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/efa.h:548]: fixed.

File: prov/efa/src/efa.h [hunk 9]
```diff
@@ -544,6 +544,34 @@ bool efa_both_support_rdma_read(struct rxr_ep *ep, struct rdm_peer *peer)
...
+ * int req packet header.
+ *
+ * There are two cases a peer need the raw address in REQ packet header:
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/efa.h:551]: fixed

File: prov/efa/src/efa.h [hunk 10]
```diff
@@ -544,6 +544,34 @@ bool efa_both_support_rdma_read(struct rxr_ep *ep, struct rdm_peer *peer)
...
+ *
+ * First, the first several packets to a peer should include raw address
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/efa.h:553]: fixed

File: prov/efa/src/efa.h [hunk 11]
```diff
@@ -544,6 +544,34 @@ bool efa_both_support_rdma_read(struct rxr_ep *ep, struct rdm_peer *peer)
...
+ * because the peer might not have ep's address in its address vector, thus
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/efa.h:554]: fixed

File: prov/efa/src/efa.h [hunk 12]
```diff
@@ -544,6 +544,34 @@ bool efa_both_support_rdma_read(struct rxr_ep *ep, struct rdm_peer *peer)
...
+ * cannot send packet back.
+ *
+ * Normally, after and endpoint received a hanshake packet from a peer, it
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/efa.h:557]: fixed

File: prov/efa/src/efa.h [hunk 13]
```diff
@@ -544,6 +544,34 @@ bool efa_both_support_rdma_read(struct rxr_ep *ep, struct rdm_peer *peer)
...
+ * can stop including raw addres in packet header. However, if the peer
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/efa.h:558]: fixed

File: prov/efa/src/efa.h [hunk 14]
```diff
@@ -544,6 +544,34 @@ bool efa_both_support_rdma_read(struct rxr_ep *ep, struct rdm_peer *peer)
...
+ * is in zero copy receive mode, the endpoint have to keep including raw
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/efa.h:559]: fixed

File: prov/efa/src/efa.h [hunk 15]
```diff
@@ -544,6 +544,34 @@ bool efa_both_support_rdma_read(struct rxr_ep *ep, struct rdm_peer *peer)
...
+ * address in header because zero copy receive requires the message prefix
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/efa.h:560]: fixed.

File: prov/efa/src/efa.h [hunk 16]
```diff
@@ -544,6 +544,34 @@ bool efa_both_support_rdma_read(struct rxr_ep *ep, struct rdm_peer *peer)
...
+ * size to remain same.
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/efa.h:561]: fixed

File: prov/efa/src/efa.h [hunk 17]
```diff
@@ -544,6 +544,34 @@ bool efa_both_support_rdma_read(struct rxr_ep *ep, struct rdm_peer *peer)
...
+ *
+ * @params[in]	peer	pointer to rdm_peer
+ * @return	a boolean indicating whether the peer need raw address header
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/efa.h:564]: fixed

File: prov/efa/src/rxr/rxr.h [hunk 18]
```diff
@@ -886,8 +886,9 @@ int rxr_endpoint(struct fid_domain *domain, struct fi_info *info,
 /* EP sub-functions */
 void rxr_ep_progress(struct util_ep *util_ep);
 void rxr_ep_progress_internal(struct rxr_ep *rxr_ep);
-int rxr_ep_post_buf(struct rxr_ep *ep, const struct fi_msg *posted_recv,
-		    uint64_t flags, enum rxr_lower_ep_type lower_ep);
+
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr.h:889]: fixed

File: prov/efa/src/rxr/rxr_msg.c [hunk 19]
```diff
@@ -1047,13 +1047,14 @@ ssize_t rxr_msg_generic_recv(struct fid_ep *ep, const struct fi_msg *msg,
 		goto out;
 	}
 
-	if (op == ofi_op_tagged)
+	if (rxr_ep->use_zcpy_rx) {
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_msg.c:1050]: Will split this change in a separate commit

File: prov/efa/src/rxr/rxr_msg.c (hunk 7 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_msg.c]: fixed

File: prov/efa/src/rxr/rxr_ep.c [hunk 20]
```diff
@@ -275,17 +274,14 @@ void rxr_tx_entry_init(struct rxr_ep *ep, struct rxr_tx_entry *tx_entry,
...
-	/*
-	 * The prefix is currently not used by the sender, but needs to be
-	 * accounted for when copying the payload into the bounce-buffer.
-	 */
-	if (ep->use_zcpy_rx) {
-		assert(tx_entry->iov[0].iov_len >= sizeof(struct rxr_pkt_entry) + sizeof(struct rxr_eager_msgrtm_hdr));
-		tx_entry->iov[0].iov_base = (char *)tx_entry->iov[0].iov_base
-					     + sizeof(struct rxr_pkt_entry)
-					     + sizeof(struct rxr_eager_msgrtm_hdr);
+	if (ep->msg_prefix_size) {
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:372]: ep->msg_prefix_size is always being set, but the value is not always > 0.

Only when application pass a non-zero msg_prefix_size, will this path be used.

File: prov/efa/src/rxr/rxr_ep.c (hunk 3 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:304]: fixed

File: prov/efa/src/rxr/rxr_ep.c (hunk 5 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c]: fixed

**Comment by wzamazon** [comment]
@wckzhang I believe I addressed all your comments, except for one. For that I need some time

**Comment by wckzhang** [comment]
> @wckzhang I believe I addressed all your comments, except for one. For that I need some time

Which one?

**Comment by wzamazon** [comment]
The commit that split `rxr_ep_post_buf` into 2 functions also changed some behavior, I will split that commit into 2.

File: prov/efa/src/rxr/rxr_msg.c (hunk 19 above)
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_msg.c:1045]: Added a separate commit for this change.

File: prov/efa/src/rxr/rxr_pkt_cmd.c [hunk 21]
```diff
@@ -966,6 +918,78 @@ void rxr_pkt_handle_recv_completion(struct rxr_ep *ep,
...
+		rxr_pkt_post_handshake_or_queue(ep, peer);
+
+	if (peer->is_local) {
+		assert(ep->use_shm);
+		ep->posted_bufs_shm--;
+	} else {
+		ep->posted_bufs_efa--;
+	}
+
+	zcpy_rx_entry = NULL;
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_cmd.c]: Done

**Comment by wzamazon** [comment]
@wckzhang I added the one commit as mentioned

File: prov/efa/src/efa.h [hunk 22]
```diff
@@ -544,6 +544,35 @@ bool efa_both_support_rdma_read(struct rxr_ep *ep, struct rdm_peer *peer)
...
+ * @brief determines whether a peer needs the endpoint to include
+ * raw address int the req packet header.
+ *
+ * There are two cases a peer need the raw address in REQ packet header:
+ *
+ * 1. the initial packets to a peer should include the raw address,
+ * because the peer might not have ep's address in its address vector
+ * causing the peer to be unable to send packet back. Normally, after
+ * an endpoint received a hanshake packet from a peer, it can stop
+ * including raw addres in packet header.
```
Comment (wckzhang) [inline prov/efa/src/efa.h:557]: addres -> address

File: prov/efa/src/rxr/rxr_msg.c [hunk 23]
```diff
@@ -1047,15 +1047,14 @@ ssize_t rxr_msg_generic_recv(struct fid_ep *ep, const struct fi_msg *msg,
 		goto out;
 	}
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_msg.c:1043]: Couple typos in the commit message for ```prov/efa: change the way pkt_entry matching with rx_entry in zero copy
```

communcations -> communications

laster remove -> later removed

**Review by wckzhang — APPROVED** [review]

File: prov/efa/src/rxr/rxr_msg.c (hunk 23 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_msg.c:1043]: Thank you! fixed

File: prov/efa/src/efa.h (hunk 22 above)
Comment (wzamazon) [reply prov/efa/src/efa.h]: fixed

**Comment by wzamazon** [comment]
waiting for CI to finish.

**Review by bwbarrett — APPROVED** [review]

File: prov/efa/src/rxr/rxr_ep.c [hunk 24]
```diff
@@ -758,6 +758,9 @@ void rxr_ep_set_features(struct rxr_ep *ep)
 		ep->features[0] |= RXR_REQ_FEATURE_RDMA_READ;
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:848]: added commit message.

**Comment by wzamazon** [comment]
bot:aws:retest

**Pull request merged**